### ISSUE TITLE
imtcp: add input() parameters for NSD driver options

### ISSUE THREAD

**Issue by skupkosk** [rsyslog/rsyslog#3727]
================== Maintainer edit ==================
See below: this actually is about **adding additional input parameters** rather then moving parameters. Relevant comment with detail info: https://github.com/rsyslog/rsyslog/issues/3727#issuecomment-524750746
================ End Maintainer edit ==================

### Expected behavior
Be able to use imtcp module for both, clean and TLS inputs. The similar way it is used on omfwd actions. Every omfwd action can use their own StreamDriver options.

### Actual behavior
The module is loaded once with NSD preloaded without possibility to change in input() calls.
At the moment the only workaround is the availability of imptcp which can be used for plain TCP.

**Comment by davidelang** [comment]
we don't want to remove the tls settings from the module() call for backwards 
compatibility reasons

But we do need to make it possible to set them on the input() call. There are 
just too may reasons to need to have different settings for differen inputs 
(encrypted, not encrypted, different certs on different inputs, different auth 
modes on different ports, etc)

David Lang

**Comment by davidelang** [comment]
another possible options would be to allow a module to be loaded multiple times, 
with different names (the internals would have to end up in their own namespaces 
to avoid conflicts)

David Lang

**Comment by skupkosk** [comment]
Would prefer the way of not needing loading module multiple times.

The side-or-main motivation could be the ability to throw away one of imtcp or imptcp modules which are just doing the same.

**Comment by davidelang** [comment]
as I understood it, the 'p' in imptcp was origionally for 'parallel', in that it 
allowed more than one thread to work on the input from one connection. This is a 
big performance win if you have a lot of logs from one source, but you can't do 
encryption in parallel like that, so imtcp will need to stay around.

I agree that loading the module multiple times would not be ideal, but if it 
could be done with less development effort, it may be a win.

David Lang

**Comment by rgerhards** [comment]
Loading a module multiple times is not possible nor desirable - it would require a complete redesign. This should be changed as follows:

* make module parameters act as defaults
* add input parameters

Actually, I thought this was already done, but it doesn't look so. Out of my head I do not know if it is a smaller or larger task. Tentatively assigning milestone, let's see if someone has time to look at it.

**Comment by rgerhards** [comment]
I had a look at the code. It's unfortunately not a small job. The server is currently written in a way that assigns the parameter at the global server basis, not the individual listener. I'll further evaluate what needs to be changed to support that.

**Comment by rgerhards** [comment]
We unfortunately need to changes this through all layers down to the driver level. Another problem is that some properties even stem back from `glbl` object settings. I guess this is a left-over from old style configuration language.

An idea is to introduce a "listener parameter block" which is consistently passed down through the layers and contains all information ever necessary (and thus can also easily be extended). This could be done in a multi-step approach:

1. introduce the parameter block for params currently passed down via function arguments
2. shuffle the TLS settings to the param block, but still use the global valriables
3. refactor the rest of the system so that the TLS params are indeed listener specific

This iterative approach also ensures that we can apply proper testing and do not need to refactor too much of the code at the same time.
